fix dynamic shapes for kwargs

avikchaudhuri opened the pull request:
Summary:
In this PR we change how `dynamic_shapes` map to the top-level structure of `args` and `kwargs`.

Previously, we would match `dynamic_shapes` to the input signature of a module, using `inspect.Signature.bind`; instead, now we match it with the structure of `args` and `kwargs`.

This has some desirable consequences:
1. Specifying dynamic shapes of variadic args and kwargs becomes less confusing. Previously we would have to have additional structure in `dynamic_shapes` to match the "tuple" / "dict" nature of `*args` / `**kwargs`, now we can directly map inputs.
2. We do not suffer from brittleness when modules are wrapped, typically with `*args, **kwargs`. If your inputs work, your dynamic shapes should also work.

In the new regime, you still have a choice on whether to specify `dynamic_shapes` as a tuple/list or a dict.
* As a tuple/list, you're basically associating with `*args, *kwargs.values()`
* As a dict, you (obviously) must use the same names as `kwargs` at least. Moreover, to avoid silly mistakes you're also expected to use names in the input signature for `args`, as long as it is not variadic—otherwise you can use whatever names you want. This last decision is a subjective choice, striking a balance between error reporting and flexibility.

Fixes #150022
Fixes #150371

Test Plan: added / fixed tests

Differential Revision: D72350333


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

titaiwangms commented:
Does this PR also cover wrapped model(*args, **kwargs):  for example, https://github.com/pytorch/pytorch/issues/148713?

avikchaudhuri commented:
> Does this PR also cover wrapped model(*args, **kwargs): for example, #148713?

It should, I'll add a test to illustrate.

tugsbayasgalan commented on torch/export/dynamic_shapes.py:
```diff
@@ -623,22 +623,92 @@ def _compare(tree, dynamic_shapes, path):
 
 
+def _signature(f):
+    if isinstance(f, ExportedProgram):
```
When is it possible to have ExportedProgram as input to export?

avikchaudhuri replied:
I think if you re-export it? Not sure, but this check exists before this PR.

Skylion007 commented on torch/export/dynamic_shapes.py:
```diff
@@ -623,22 +623,94 @@ def _compare(tree, dynamic_shapes, path):
+    )
+
+
+def _split_positional_args(f, args) -> tuple[dict[str, Any], dict[str, Any]]:
```
Is ParamSpec useful here?

avikchaudhuri committed:
fix dynamic shapes for kwargs (#150583)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/150583

In this PR we change how `dynamic_shapes` map to the top-level structure of `args` and `kwargs`.

Previously, we would match `dynamic_shapes` to the input signature of a module, using `inspect.Signature.bind`; instead, now we match it with the structure of `args` and `kwargs`.

This has some desirable consequences:
1. Specifying dynamic shapes of variadic args and kwargs become less confusing. Previously we would have to have additional structure in `dynamic_shapes` to match the "tuple" / "dict" nature of `*args` / `**kwargs`, now we can directly map inputs.
2. We do not suffer from brittleness when modules are wrapped, typically with `*args, **kwargs`. If your inputs work, your dynamic shapes should also work.

In the new regime, you still have a choice on whether to specify `dynamic_shapes` as a tuple/list or a dict.
* As a tuple/list, you're basically associating with `*args, *kwargs.values()`
* As a dict, you (obviously) must use the same names as `kwargs` at least. Moreover, to avoid silly mistakes you're also expected to use names in the input signature for `args`, as long as it is not variadic—otherwise you can use whatever names you want. This last decision is a subjective choice, striking a balance between error reporting and flexibility.

Fixes #150022
Fixes #150371

Test Plan: added / fixed tests

Reviewed By: tugsbayasgalan

Differential Revision: D72350333
